File: day7/ex2.py
with open("test_data.txt", "r") as fd:
    data = list(map(int, fd.read().strip().split(",")))


def cost(f, t):
    diff = abs(f - t) + 1
    return int(diff * (diff - 1) / 2)


mean = round(sum(data) / len(data))
answer = min(
    sum(cost(p, mean + i) for p in data)
    for i in (-1, 0, 1)
)

print(answer)

File: day7/ex2.zig
const std = @import("std");
const stdout = std.io.getStdOut().writer();

const data = @embedFile("test_data.txt");

pub fn main() !void {
    var crabs = std.ArrayList(usize).init(std.heap.page_allocator);
    defer crabs.deinit();

    var sum: usize = 0;
    var data_it = std.mem.tokenize(u8, data, ",\n");
    while (data_it.next()) |token| {
        const crab = try std.fmt.parseUnsigned(usize, token, 10);
        try crabs.append(crab);
        sum += crab;
    }

    const mean = try std.math.divCeil(usize, sum, crabs.items.len);
    var answers = [_]usize{ 0, 0, 0 };

    for (crabs.items) |crab| {
        answers[0] += cost(crab, mean - 1);
        answers[1] += cost(crab, mean);
        answers[2] += cost(crab, mean + 1);
    }
    const answer = std.mem.min(usize, &answers);
    try stdout.print("answer: {}\n", .{answer});
}

fn cost(from: usize, to: usize) usize {
    const diff = if (from > to) from - to else to - from;
    return diff * (diff + 1) / 2;
}
